PR: https://github.com/odoo/odoo/pull/244436

From: d37cf89a6ff134988b4a7c001a97973cd95a2ea8
From: Yannick Tivisse
Date: 2026-02-14 18:04:30

Breaking data model changes scores: del:28.4 + add:0, change matches:
-    has_work_entries = fields.Boolean(compute='_compute_has_work_entries', groups="base.group_system,hr.group_hr_user")
-    _inherit = 'hr.employee.departure'
-    work_entries_warning_date = fields.Date(compute='_compute_work_entries_warning_date')
-    date_generated_from = fields.Datetime(string='Generated From', readonly=True, required=True,
-    date_generated_to = fields.Datetime(string='Generated To', readonly=True, required=True,
-    last_generation_date = fields.Date(string='Last Generation Date', readonly=True, groups="hr.group_hr_user", tracking=True)
-    name = fields.Char("Description")
-    active = fields.Boolean(default=True)
-    employee_id = fields.Many2one('hr.employee', required=True, domain="['|', ('company_id', '=', False), ('company_id', '=', company_id)]", index=True)
-    version_id = fields.Many2one('hr.version', string="Employee Record", required=True, index=True)
-    work_entry_source = fields.Selection(related='version_id.work_entry_source')
-    resource_calendar_id = fields.Many2one(related='version_id.resource_calendar_id')
-    date = fields.Date(required=True)
-    duration = fields.Float(string="Duration", default=lambda self: self.env.company.resource_calendar_id.hours_per_day)
-    work_entry_type_id = fields.Many2one(
-    display_code = fields.Char(related='work_entry_type_id.display_code')
-    shortcut_behavior = fields.Selection(related='work_entry_type_id.shortcut_behavior')
-    color = fields.Integer(related='work_entry_type_id.color', readonly=True)
-    category = fields.Selection(related='work_entry_type_id.category')
-    is_manual = fields.Boolean(help="Marks if the work entry came from manual creation or split")
-    state = fields.Selection([
-    company_id = fields.Many2one('res.company', string='Company', readonly=True, required=True,
-    conflict = fields.Boolean('Conflicts', compute='_compute_conflict', store=True)
-    department_id = fields.Many2one('hr.department', related='employee_id.department_id', store=True)
-    amount_rate = fields.Float("Pay rate")
-    country_id = fields.Many2one('res.country', related='employee_id.company_id.country_id', search='_search_country_id')
-    code = fields.Char(string="Payroll Code", required=True, help="The code is used as a reference in salary rules. Careful, changing an existing code can lead to unwanted behaviors.")
-    category = fields.Selection(
+    work_entry_type_id = fields.Many2one('hr.work.entry.type', 'Work Entry Type', store=True,
-    work_entry_type_id = fields.Many2one(

Total Changes: 841

[IMP] hr_work_entry: Refactor work entries / time off management

COM PR: https://github.com/odoo/odoo/pull/244436
ENT PR: https://github.com/odoo/enterprise/pull/104722

closes odoo/odoo#244436

Taskid: 5407736
Related: odoo/enterprise#104722
Related: odoo/upgrade#9378
Signed-off-by: Yannick Tivisse (yti) <yti@odoo.com>
Co-authored-by: romc-odoo <romc@odoo.com>
Co-authored-by: Jurgen Gjini <jugj@odoo.com>
Co-authored-by: Mahmoudk3m <kmah@odoo.com>

================================= pseudo patch: =================================

--- a/addons/hr_work_entry/models/__init__.py
+++ b/addons/hr_work_entry/models/__init__.py
@@ -1,12 +1,12 @@
 # -*- coding:utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
-from . import hr_work_entry
 from . import hr_work_entry_type
 from . import hr_user_work_entry_employee
 from . import resource_calendar
 from . import resource_calendar_attendance
 from . import resource_calendar_leaves
 from . import hr_employee
-from . import hr_employee_departure
 from . import hr_version
+from . import res_company
+from . import res_config_settings

--- a/addons/hr_work_entry/models/hr_employee.py
+++ b/addons/hr_work_entry/models/hr_employee.py
@@ -1,54 +1,18 @@
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
-from odoo import models, fields, _
-from odoo.tools import SQL
+from odoo import models, fields
 
 
 class HrEmployee(models.Model):
     _inherit = 'hr.employee'
 
-    has_work_entries = fields.Boolean(compute='_compute_has_work_entries', groups="base.group_system,hr.group_hr_user")
     work_entry_source = fields.Selection(readonly=False, related="version_id.work_entry_source", inherited=True, groups="base.group_system,hr.group_hr_manager")
     work_entry_source_calendar_invalid = fields.Boolean(related="version_id.work_entry_source_calendar_invalid", inherited=True, groups="hr.group_hr_manager")
+    external_code = fields.Char("External Code", copy=False, help="Use this code to export your data to a third party", groups="hr.group_hr_user")
 
-    def _compute_has_work_entries(self):
-        if self.ids:
-            result = dict(self.env.execute_query(SQL(
-                """ SELECT id, EXISTS(SELECT 1 FROM hr_work_entry WHERE employee_id = e.id LIMIT 1)
-                      FROM hr_employee e
-                     WHERE id in %s """,
-                tuple(self.ids),
-            )))
-        else:
-            result = {}
-
-        for employee in self:
-            employee.has_work_entries = result.get(employee._origin.id, False)
-
-    def create_version(self, values):
-        new_version = super().create_version(values)
-        new_version.update({
-            'date_generated_from': fields.Datetime.now().replace(hour=0, minute=0, second=0, microsecond=0),
-            'date_generated_to': fields.Datetime.now().replace(hour=0, minute=0, second=0, microsecond=0),
-        })
-        return new_version
-
-    def action_open_work_entries(self, initial_date=False):
-        self.ensure_one()
-        ctx = {'default_employee_id': self.id}
-        if initial_date:
-            ctx['initial_date'] = initial_date
-        return {
-            'type': 'ir.actions.act_window',
-            'name': _('%s work entries', self.display_name),
-            'view_mode': 'calendar,list,form',
-            'res_model': 'hr.work.entry',
-            'path': 'work-entries',
-            'context': ctx,
-            'domain': [('employee_id', '=', self.id)],
-        }
-
-    def generate_work_entries(self, date_start, date_stop, force=False):
+    # YTI TODO: Rename private method into _get_work_entries_vals()
+    # Public method probably to drop
+    def generate_work_entries(self, date_start, date_stop):
         date_start = fields.Date.to_date(date_start)
         date_stop = fields.Date.to_date(date_stop)
 
@@ -56,4 +20,4 @@ class HrEmployee(models.Model):
             versions = self._get_versions_with_contract_overlap_with_period(date_start, date_stop)
         else:
             versions = self._get_all_versions_with_contract_overlap_with_period(date_start, date_stop)
-        return versions.generate_work_entries(date_start, date_stop, force=force)
+        return versions.generate_work_entries(date_start, date_stop)

--- a/addons/hr_work_entry/models/hr_employee_departure.py
+++ b/None
@@ -1,44 +0,0 @@
-# Part of Odoo. See LICENSE file for full copyright and licensing details.
-
-from odoo import api, fields, models
-from odoo.fields import Domain
-
-
-class HrEmployeeDeparture(models.Model):
-    _inherit = 'hr.employee.departure'
-
-    work_entries_warning_date = fields.Date(compute='_compute_work_entries_warning_date')
-
-    def _get_future_validated_work_entries(self):
-        domain = Domain.AND([
-            Domain('state', '=', 'validated'),
-            Domain.OR([
-                ('date', '>', dep.departure_date),
-                ('employee_id', '=', dep.employee_id.id),
-            ] for dep in self),
-        ])
-        return dict(self.sudo().env['hr.work.entry']._read_group(
-            domain=domain,
-            groupby=['employee_id'],
-            aggregates=['id:recordset'],
-        ))
-
-    @api.depends('do_set_date_end', 'departure_date')
-    def _compute_work_entries_warning_date(self):
-        validated_work_entries_sudo = self._get_future_validated_work_entries()
-        self.work_entries_warning_date = False
-        for departure in self:
-            future_work_entries = validated_work_entries_sudo.get(departure.employee_id)
-            if not departure.do_set_date_end or not future_work_entries:
-                continue
-            departure.work_entries_warning_date = max(future_work_entries.mapped('date'))
-
-    def action_register(self):
-        validated_work_entries_sudo = self._get_future_validated_work_entries()
-        we_to_draft = self.env['hr.work.entry']
-        for departure in self:
-            future_work_entries = validated_work_entries_sudo.get(departure.employee_id)
-            if departure.do_set_date_end and future_work_entries:
-                we_to_draft += future_work_entries
-        we_to_draft.with_context({'departure_unlink': True}).write({'state': 'draft'})
-        return super().action_register()

--- a/addons/hr_work_entry/models/hr_version.py
+++ b/addons/hr_work_entry/models/hr_version.py
@@ -2,14 +2,12 @@
 
 import itertools
 from collections import defaultdict
-from datetime import datetime, date, time, timedelta, UTC
+from datetime import datetime, time, timedelta, UTC
 from zoneinfo import ZoneInfo
 
-from dateutil.relativedelta import relativedelta
-
 from odoo import api, fields, models, _, SUPERUSER_ID
 from odoo.exceptions import UserError
-from odoo.fields import Command, Domain
+from odoo.fields import Domain
 from odoo.tools import ormcache, float_is_zero
 from odoo.tools.intervals import Intervals
 
@@ -17,13 +15,6 @@ from odoo.tools.intervals import Intervals
 class HrVersion(models.Model):
     _inherit = 'hr.version'
 
-    date_generated_from = fields.Datetime(string='Generated From', readonly=True, required=True,
-        default=lambda self: datetime.now().replace(hour=0, minute=0, second=0, microsecond=0),
-        groups="hr.group_hr_user", tracking=True)
-    date_generated_to = fields.Datetime(string='Generated To', readonly=True, required=True,
-        default=lambda self: datetime.now().replace(hour=0, minute=0, second=0, microsecond=0),
-        groups="hr.group_hr_user", tracking=True)
-    last_generation_date = fields.Date(string='Last Generation Date', readonly=True, groups="hr.group_hr_user", tracking=True)
     work_entry_source = fields.Selection([('calendar', 'Working Schedule')], required=True, default='calendar', tracking=True, help='''
         Defines the source for work entries generation
 
@@ -41,9 +32,10 @@ class HrVersion(models.Model):
         for version in self:
             version.work_entry_source_calendar_invalid = version.work_entry_source == 'calendar' and not version.resource_calendar_id
 
+    # YTI Break ormcache + select country attendance entry type
     @ormcache()
     def _get_default_work_entry_type_id(self):
-        attendance = self.env.ref('hr_work_entry.work_entry_type_attendance', raise_if_not_found=False)
+        attendance = self.env.ref('hr_work_entry.generic_work_entry_type_attendance', raise_if_not_found=False)
         return attendance.id if attendance else False
 
     def _get_leave_work_entry_type_dates(self, leave, date_from, date_to, employee):
@@ -56,7 +48,7 @@ class HrVersion(models.Model):
     def _get_more_vals_attendance_interval(self, interval):
         return []
 
-    # Is used to add more values, for example leave_id (in hr_work_entry_holidays)
+    # Is used to add more values, for example leave_id (in hr_holidays)
     def _get_more_vals_leave_interval(self, interval, leaves):
         return []
 
@@ -74,7 +66,7 @@ class HrVersion(models.Model):
                 interval_start = interval[0].astimezone(UTC).replace(tzinfo=None)
                 interval_stop = interval[1].astimezone(UTC).replace(tzinfo=None)
                 return self._get_leave_work_entry_type_dates(leave[2], interval_start, interval_stop, self.employee_id)
-        return self.env.ref('hr_work_entry.work_entry_type_leave')
+        return self.env.ref('hr_work_entry.generic_work_entry_type_leave')
 
     def _get_sub_leave_domain(self):
         return Domain('calendar_id', 'in', [False] + self.resource_calendar_id.ids)
@@ -94,17 +86,13 @@ class HrVersion(models.Model):
     def _get_attendance_intervals(self, start_dt, end_dt):
         assert start_dt.tzinfo and end_dt.tzinfo, "function expects localized date"
         # {resource: intervals}
-        versions_by_calendar = defaultdict(lambda: self.env['hr.version'])
-        for version in self:
-            if version.work_entry_source != 'calendar':
-                continue
-            versions_by_calendar[version.resource_calendar_id] |= version
+        versions_with_calendar_work_entry_source = self.filtered(lambda version: version.work_entry_source == 'calendar')
         result = dict()
-        for calendar, versions in versions_by_calendar.items():
+        for calendar, versions in versions_with_calendar_work_entry_source.grouped('resource_calendar_id').items():
             fully_flex_versions = versions.filtered('is_fully_flexible')
             for version in fully_flex_versions:
                 result.update({version.employee_id.resource_id.id: Intervals([(start_dt, end_dt, self.env['resource.calendar.attendance'])])})
-            remaining_versions = versions - fully_flex_versions
+            remaining_versions = (versions - fully_flex_versions).with_prefetch()
             resources_per_tz = remaining_versions._get_resources_per_tz()
             if remaining_versions:
                 result.update(calendar._attendance_intervals_batch(
@@ -139,10 +127,10 @@ class HrVersion(models.Model):
             vals += [dict([
                       ('date_start', interval[0].astimezone(UTC).replace(tzinfo=None)),
                       ('date_stop', interval[1].astimezone(UTC).replace(tzinfo=None)),
-                      ('work_entry_type_id', work_entry_type.id),
-                      ('employee_id', employee.id),
-                      ('version_id', self.id),
-                      ('company_id', self.company_id.id),
+                      ('work_entry_type_id', work_entry_type),
+                      ('employee_id', employee),
+                      ('version_id', self),
+                      ('company_id', self.company_id),
                   ] + self._get_more_vals_attendance_interval(interval))]
         return vals
 
@@ -152,13 +140,13 @@ class HrVersion(models.Model):
         version_vals = []
         bypassing_work_entry_type_codes = self._get_bypassing_work_entry_type_codes()
 
-        attendances_by_resource = self.sudo()._get_attendance_intervals(start_dt, end_dt)
+        expected_attendances_by_resource = self.sudo()._get_attendance_intervals(start_dt, end_dt)
 
         resource_calendar_leaves = self._get_resource_calendar_leaves(start_dt, end_dt)
         # {resource: resource_calendar_leaves}
-        leaves_by_resource = defaultdict(lambda: self.env['resource.calendar.leaves'])
+        all_leaves_by_resource = defaultdict(lambda: self.env['resource.calendar.leaves'])
         for leave in resource_calendar_leaves:
-            leaves_by_resource[leave.resource_id.id] |= leave
+            all_leaves_by_resource[leave.resource_id.id] |= leave
 
         tz_dates = {}
         for version in self:
@@ -166,14 +154,14 @@ class HrVersion(models.Model):
             calendar = version.resource_calendar_id
             resource = employee.resource_id
             tz = ZoneInfo(version._get_tz())
-            attendances = attendances_by_resource[resource.id]
+            expected_attendances = expected_attendances_by_resource[resource.id]
 
             # Other calendars: In case the employee has declared time off in another calendar
             # Example: Take a time off, then a credit time.
             resources_list = [self.env['resource.resource'], resource]
             leave_result = defaultdict(list)
             work_result = defaultdict(list)
-            for leave in itertools.chain(leaves_by_resource[False], leaves_by_resource[resource.id]):
+            for leave in itertools.chain(all_leaves_by_resource[False], all_leaves_by_resource[resource.id]):
                 for resource in resources_list:
                     # Global time off is not for this calendar, can happen with multiple calendars in self
                     if resource and leave.calendar_id and leave.calendar_id != calendar and not leave.resource_id:
@@ -194,63 +182,42 @@ class HrVersion(models.Model):
                     leave_start_dt = max(start, dt0)
                     leave_end_dt = min(end, dt1)
                     leave_interval = (leave_start_dt, leave_end_dt, leave)
-                    leave_interval = version._get_valid_leave_intervals(attendances, leave_interval)
+                    leave_interval = version._get_valid_leave_intervals(expected_attendances, leave_interval)
                     if leave_interval:
-                        if leave.time_type == 'leave':
+                        if leave.count_as == 'absence':
                             leave_result[resource.id] += leave_interval
                         else:
                             work_result[resource.id] += leave_interval
-            mapped_leaves = {r.id: Intervals(leave_result[r.id], keep_distinct=True) for r in resources_list}
-            mapped_worked_leaves = {r.id: Intervals(work_result[r.id], keep_distinct=True) for r in resources_list}
+            leaves_by_resource = {r.id: Intervals(leave_result[r.id], keep_distinct=True) for r in resources_list}
+            worked_leaves_by_resource = {r.id: Intervals(work_result[r.id], keep_distinct=True) for r in resources_list}
 
-            leaves = mapped_leaves[resource.id]
-            worked_leaves = mapped_worked_leaves[resource.id]
+            leaves = leaves_by_resource[resource.id]
+            worked_leaves = worked_leaves_by_resource[resource.id]
 
-            real_attendances = attendances - leaves - worked_leaves
+            real_attendances = expected_attendances - leaves - worked_leaves
             if version.is_fully_flexible:
                 real_leaves = leaves
-                real_worked_leaves = worked_leaves
             elif version.is_flexible:
                 # Flexible hours case
                 # For multi day leaves, we want them to occupy the virtual working schedule 12 AM to average working days
                 # For one day leaves, we want them to occupy exactly the time it was taken, for a time off in days
                 # this will mean the virtual schedule and for time off in hours the chosen hours
-                one_day_leaves = Intervals([l for l in leaves if l[0].date() == l[1].date()], keep_distinct=True)
-                one_day_worked_leaves = Intervals([l for l in worked_leaves if l[0].date() == l[1].date()], keep_distinct=True)
+                one_day_leaves = Intervals([l for l in leaves if l[0].astimezone(tz).date() == l[1].astimezone(tz).date()], keep_distinct=True)
                 multi_day_leaves = leaves - one_day_leaves
-                multi_day_worked_leaves = worked_leaves - one_day_worked_leaves
                 resources_per_tz = version._get_resources_per_tz()
                 static_attendances = calendar._attendance_intervals_batch(
                     start_dt, end_dt, resources_per_tz=resources_per_tz)[resource.id]
                 real_leaves = (static_attendances & multi_day_leaves) | one_day_leaves
-                real_worked_leaves = (static_attendances & multi_day_worked_leaves) | one_day_worked_leaves
-
             elif version.has_static_work_entries() or not leaves:
-                # Empty leaves means empty real_leaves
-                real_worked_leaves = attendances - real_attendances - leaves
-                real_leaves = attendances - real_attendances - real_worked_leaves
+                real_leaves = expected_attendances & leaves
             else:
-                # In the case of attendance based versions use regular attendances to generate leave intervals
                 resources_per_tz = version._get_resources_per_tz()
                 static_attendances = calendar._attendance_intervals_batch(
                     start_dt, end_dt, resources_per_tz=resources_per_tz)[resource.id]
                 real_leaves = static_attendances & leaves
-                real_worked_leaves = static_attendances & worked_leaves
-
-            real_attendances = self._get_real_attendances(attendances, leaves, worked_leaves)
-
-            if not version.has_static_work_entries():
-                # An attendance based version might have an invalid planning, by definition it may not happen with
-                # static work entries.
-                # Creating overlapping slots for example might lead to a single work entry.
-                # In that case we still create both work entries to indicate a problem (conflicting W E).
-                split_attendances = []
-                for attendance in real_attendances:
-                    if attendance[2] and len(attendance[2]) > 1:
-                        split_attendances += [(attendance[0], attendance[1], a) for a in attendance[2]]
-                    else:
-                        split_attendances += [attendance]
-                real_attendances = split_attendances
+
+            real_worked_leaves = worked_leaves - real_leaves
+            real_attendances = self._get_real_attendances(expected_attendances, leaves, worked_leaves)
 
             # A leave period can be linked to several resource.calendar.leave
             split_leaves = []
@@ -278,11 +245,10 @@ class HrVersion(models.Model):
                 version_vals += [dict([
                     ('date_start', interval[0].astimezone(UTC).replace(tzinfo=None)),
                     ('date_stop', interval[1].astimezone(UTC).replace(tzinfo=None)),
-                    ('work_entry_type_id', work_entry_type.id),
-                    ('employee_id', employee.id),
-                    ('version_id', version.id),
-                    ('company_id', version.company_id.id),
-                    ('state', 'draft'),
+                    ('work_entry_type_id', work_entry_type),
+                    ('employee_id', employee),
+                    ('version_id', version),
+                    ('company_id', version.company_id),
                 ] + version._get_more_vals_leave_interval(interval, worked_leaves))]
 
             leaves_over_attendances = Intervals(leaves, keep_distinct=True) & real_leaves
@@ -306,10 +272,10 @@ class HrVersion(models.Model):
                     version_vals += [dict([
                         ('date_start', interval_start),
                         ('date_stop', interval_stop),
-                        ('work_entry_type_id', leave_entry_type.id),
-                        ('employee_id', employee.id),
-                        ('company_id', version.company_id.id),
-                        ('version_id', version.id),
+                        ('work_entry_type_id', leave_entry_type),
+                        ('employee_id', employee),
+                        ('company_id', version.company_id),
+                        ('version_id', version),
                     ] + version._get_more_vals_leave_interval(interval, interval_leaves))]
         return version_vals
 
@@ -335,29 +301,6 @@ class HrVersion(models.Model):
                     date_start.replace(tzinfo=tz),
                     date_stop.replace(tzinfo=tz))
 
-        # {version_id: ([dates_start], [dates_stop])}
-        mapped_version_dates = defaultdict(lambda: ([], []))
-        for x in version_vals:
-            mapped_version_dates[x['version_id']][0].append(x['date_start'])
-            mapped_version_dates[x['version_id']][1].append(x['date_stop'])
-
-        for version in self:
-            # If we generate work_entries which exceeds date_start or date_stop, we change boundaries on version
-            if version_vals:
-                # Handle empty work entries for certain versions, could happen on an attendance based version
-                # NOTE: this does not handle date_stop or date_start not being present in vals
-                dates_stop = mapped_version_dates[version.id][1]
-                if dates_stop:
-                    date_stop_max = max(dates_stop)
-                    if date_stop_max > version.date_generated_to:
-                        version.date_generated_to = date_stop_max
-
-                dates_start = mapped_version_dates[version.id][0]
-                if dates_start:
-                    date_start_min = min(dates_start)
-                    if date_start_min < version.date_generated_from:
-                        version.date_generated_from = date_start_min
-
         return version_vals
 
     def has_static_work_entries(self):
@@ -366,7 +309,7 @@ class HrVersion(models.Model):
         self.ensure_one()
         return self.work_entry_source == 'calendar'
 
-    def generate_work_entries(self, date_start, date_stop, force=False):
+    def generate_work_entries(self, date_start, date_stop):
         # Generate work entries between 2 dates (datetime.date)
         # To correctly englobe the period, the start and end periods are converted
         # using the calendar timezone.
@@ -382,16 +325,16 @@ class HrVersion(models.Model):
                 version.company_id,
                 version.tz or version.employee_id.user_id.tz,
             ] += version
-        new_work_entries = self.env['hr.work.entry']
+        new_work_entries = []
         for (company, version_tz), versions in versions_by_company_tz.items():
             tz = ZoneInfo(version_tz) if version_tz else UTC
             date_start_tz = date_start.replace(tzinfo=tz).astimezone(UTC).replace(tzinfo=None)
             date_stop_tz = date_stop.replace(tzinfo=tz).astimezone(UTC).replace(tzinfo=None)
             new_work_entries += versions.with_user(SUPERUSER_ID).with_company(company)._generate_work_entries(
-                date_start_tz, date_stop_tz, force=force)
+                date_start_tz, date_stop_tz)
         return new_work_entries
 
-    def _generate_work_entries(self, date_start, date_stop, force=False):
+    def _generate_work_entries(self, date_start, date_stop):
         # Generate work entries between 2 dates (datetime.datetime)
         # This method considers that the dates are correctly localized
         # based on the target timezone
@@ -399,18 +342,8 @@ class HrVersion(models.Model):
         assert isinstance(date_stop, datetime)
         self = self.with_context(tracking_disable=True)  # noqa: PLW0642
         vals_list = []
-        self.write({'last_generation_date': fields.Date.today()})
 
         intervals_to_generate = defaultdict(lambda: self.env['hr.version'])
-        # In case the date_generated_from == date_generated_to, move it to the date_start to
-        # avoid trying to generate several months/years of history for old versions for which
-        # we've never generated the work entries.
-        self.filtered(lambda c: c.date_generated_from == c.date_generated_to).write({
-            'date_generated_from': date_start,
-            'date_generated_to': date_start,
-        })
-        domain_to_nullify = Domain(False)
-        work_entry_null_vals = {field: False for field in self.env["hr.work.entry.regeneration.wizard"]._work_entry_fields_to_nullify()}
 
         for tz, versions in self.grouped("tz").items():
             tz = ZoneInfo(tz) if tz else UTC
@@ -418,54 +351,23 @@ class HrVersion(models.Model):
                 version_start = fields.Datetime.to_datetime(version.date_start).replace(tzinfo=tz).astimezone(UTC).replace(tzinfo=None)
                 version_stop = datetime.combine(fields.Datetime.to_datetime(version.date_end or date_stop),
                                                  datetime.max.time()).replace(tzinfo=tz).astimezone(UTC).replace(tzinfo=None)
-                if version_stop < date_stop:
-                    if version.date_generated_from != version.date_generated_to:
-                        domain_to_nullify |= Domain([
-                            ('version_id', '=', version.id),
-                            ('date', '>', version_stop.astimezone(tz)),
-                            ('date', '<=', date_stop.astimezone(tz)),
-                            ('state', '!=', 'validated'),
-                        ])
+                if version_stop < date_start:
+                    continue
                 if date_start > version_stop or date_stop < version_start:
                     continue
                 date_start_work_entries = max(date_start, version_start)
                 date_stop_work_entries = min(date_stop, version_stop)
-                if force:
-                    domain_to_nullify |= Domain([
-                        ('version_id', '=', version.id),
-                        ('date', '>=', date_start_work_entries.astimezone(tz)),
-                        ('date', '<=', date_stop_work_entries.astimezone(tz)),
-                        ('state', '!=', 'validated'),
-                    ])
-                    intervals_to_generate[date_start_work_entries, date_stop_work_entries] |= version
-                    continue
-
-                # For each version, we found each interval we must generate
-                # In some cases we do not want to set the generated dates beforehand, since attendance based work entries
-                #  is more dynamic, we want to update the dates within the _get_work_entries_values function
-                last_generated_from = min(version.date_generated_from, version_stop)
-                if last_generated_from > date_start_work_entries:
-                    version.date_generated_from = date_start_work_entries
-                    intervals_to_generate[date_start_work_entries, last_generated_from] |= version
-
-                last_generated_to = max(version.date_generated_to, version_start)
-                if last_generated_to < date_stop_work_entries:
-                    version.date_generated_to = date_stop_work_entries
-                    intervals_to_generate[last_generated_to, date_stop_work_entries] |= version
+                intervals_to_generate[date_start_work_entries, date_stop_work_entries] |= version
 
         for interval, versions in intervals_to_generate.items():
             date_from, date_to = interval
             vals_list.extend(versions._get_work_entries_values(date_from, date_to))
 
-        if domain_to_nullify != Domain.FALSE:
-            work_entries_to_nullify = self.env['hr.work.entry'].search(domain_to_nullify)
-            work_entries_to_nullify.write(work_entry_null_vals)
-
         if not vals_list:
-            return self.env['hr.work.entry']
+            return vals_list
 
         vals_list = self._generate_work_entries_postprocess(vals_list)
-        return self.env['hr.work.entry'].create(vals_list)
+        return vals_list
 
     @api.model
     def _get_work_entry_source_fields(self):
@@ -485,8 +387,8 @@ class HrVersion(models.Model):
     def _generate_work_entries_postprocess_adapt_to_calendar(self, vals):
         if 'work_entry_type_id' not in vals:
             return False
-        work_entry_type = self.env['hr.work.entry.type'].browse(vals['work_entry_type_id'])
-        return work_entry_type.category == 'absence'
+        work_entry_type = vals['work_entry_type_id']
+        return work_entry_type.count_as == 'absence'
 
     @api.model
     def _generate_work_entries_postprocess(self, vals_list):
@@ -497,15 +399,14 @@ class HrVersion(models.Model):
         cached_periods = defaultdict(float)
         tz_by_version = {}
 
-        def _get_tz(version_id):
-            if version_id in tz_by_version:
-                return tz_by_version[version_id]
-            version = self.env['hr.version'].browse(version_id)
+        def _get_tz(version):
+            if version in tz_by_version:
+                return tz_by_version[version]
             tz = version._get_tz()
             if not tz:
                 raise UserError(_('Missing timezone for work entries generation.'))
             tz = ZoneInfo(tz)
-            tz_by_version[version_id] = tz
+            tz_by_version[version] = tz
             return tz
 
         new_vals_list = []
@@ -562,7 +463,7 @@ class HrVersion(models.Model):
                     cached_periods[date_start, date_stop] = duration
                     vals['duration'] = duration
                 continue
-            version = self.env['hr.version'].browse(vals['version_id'])
+            version = vals['version_id']
             calendar = version.resource_calendar_id
             if not calendar and not version.hours_per_week and not version.hours_per_day:
                 vals['date'] = date_start.astimezone(tz).date()
@@ -582,12 +483,12 @@ class HrVersion(models.Model):
             if 'duration' not in vals:
                 date_start = vals['date_start']
                 date_stop = vals['date_stop']
-                version = self.env['hr.version'].browse(vals['version_id'])
+                version = vals['version_id']
                 calendar = version.resource_calendar_id
                 hours_per_week = version.hours_per_week
                 hours_per_day = version.hours_per_day
                 employee = version.employee_id
-                tz = _get_tz(vals['version_id'])
+                tz = _get_tz(version)
                 vals['date'] = date_start.astimezone(tz).date()
                 vals['duration'] = mapped_version_data[date_start, date_stop][calendar][employee.id]['hours'] if calendar or hours_per_week or hours_per_day else 0.0
             vals.pop('date_start', False)
@@ -610,112 +511,7 @@ class HrVersion(models.Model):
                 source_fields = self._get_work_entry_source_fields()
                 for field in source_fields:
                     if field in merged_vals[key] and field in vals:
-                        merged_vals[key][field] += vals[field]
+                        merged_vals[key][field] |= vals[field]
             else:
                 merged_vals[key] = vals.copy()
         return list(merged_vals.values())
-
-    def _remove_work_entries(self):
-        ''' Remove all work_entries that are outside contract period (function used after writing new start or/and end date) '''
-        all_we_to_unlink = self.env['hr.work.entry']
-        for version in self:
-            date_start = fields.Datetime.to_datetime(version.date_start)
-            if version.date_generated_from < date_start:
-                we_to_remove = self.env['hr.work.entry'].search([('date', '<', date_start), ('version_id', '=', version.id)])
-                if we_to_remove:
-                    version.date_generated_from = date_start
-                    all_we_to_unlink |= we_to_remove
-            if not version.date_end:
-                continue
-            date_end = datetime.combine(version.date_end, datetime.max.time())
-            if version.date_generated_to > date_end:
-                we_to_remove = self.env['hr.work.entry'].search([('date', '>', date_end), ('version_id', '=', version.id)])
-                if we_to_remove:
-                    version.date_generated_to = date_end
-                    all_we_to_unlink |= we_to_remove
-        all_we_to_unlink.unlink()
-
-    def _cancel_work_entries(self):
-        if not self:
-            return
-        domains = []
-        for version in self:
-            date_start = fields.Datetime.to_datetime(version.sudo().date_start)
-            version_domain = Domain([
-                ('version_id', '=', version.id),
-                ('date', '>=', date_start),
-            ])
-            if version.sudo().date_end:
-                date_end = datetime.combine(version.sudo().date_end, datetime.max.time())
-                version_domain &= Domain('date', '<=', date_end)
-            domains.append(version_domain)
-        domain = Domain.OR(domains) & Domain('state', '!=', 'validated')
-        work_entries = self.env['hr.work.entry'].sudo().search(domain)
-        if work_entries:
-            work_entries.unlink()
-
-    def write(self, vals):
-        result = super().write(vals)
-        if self.env.context.get('salary_simulation'):
-            return result
-        if vals.get('contract_date_end') or vals.get('contract_date_start') or vals.get('date_version'):
-            self.sudo()._remove_work_entries()
-        dependent_fields = self._get_fields_that_recompute_we()
-        if any(key in dependent_fields for key in vals):
-            for version_sudo in self.sudo():
-                date_from = max(version_sudo.date_start, version_sudo.date_generated_from.date())
-                date_to = min(version_sudo.date_end or date.max, version_sudo.date_generated_to.date())
-                if date_from != date_to and self.employee_id:
-                    version_sudo._recompute_work_entries(date_from, date_to)
-        return result
-
-    def unlink(self):
-        self._cancel_work_entries()
-        return super().unlink()
-
-    def _recompute_work_entries(self, date_from, date_to):
-        self.ensure_one()
-        if self.employee_id:
-            wizard = self.env['hr.work.entry.regeneration.wizard'].create({
-                'employee_ids': [Command.set(self.employee_id.ids)],
-                'date_from': date_from,
-                'date_to': date_to,
-            })
-            wizard.with_context(work_entry_skip_validation=True, active_test=False).regenerate_work_entries()
-
-    def _get_fields_that_recompute_we(self):
-        # Returns the fields that should recompute the work entries
-        return ['resource_calendar_id', 'work_entry_source']
-
-    @api.model
-    def _cron_generate_missing_work_entries(self):
-        # retrieve versions for the current month
-        today = fields.Date.today()
-        start = datetime.combine(today + relativedelta(day=1), time.min)
-        stop = datetime.combine(today + relativedelta(months=1, day=31), time.max)
-        all_versions = self.env['hr.employee']._get_all_versions_with_contract_overlap_with_period(start.date(), stop.date())
-        # determine versions to do (the ones whose generated dates have open periods this month)
-        versions_todo = all_versions.filtered(
-            lambda v:
-            (v.date_generated_from > start or v.date_generated_to < stop) and
-            (not v.last_generation_date or v.last_generation_date < today))
-        if not versions_todo:
-            return
-        version_todo_count = len(versions_todo)
-        # Filter versions by company, work entries generation is not supposed to be called on
-        # versions from differents companies, as we will retrieve the resource.calendar.leave
-        # and we don't want to mix everything up. The other versions will be treated when the
-        # cron is re-triggered
-        versions_todo = versions_todo.filtered(lambda v: v.company_id == versions_todo[0].company_id)
-        # generate a batch of work entries
-        BATCH_SIZE = 100
-        # Since attendance based are more volatile for their work entries generation
-        # it can happen that the date_generated_from and date_generated_to fields are not
-        # pushed to start and stop
-        # It is more interesting for batching to process statically generated work entries first
-        # since we get benefits from having multiple versions on the same calendar
-        versions_todo = versions_todo.sorted(key=lambda v: 1 if v.has_static_work_entries() else 100)
-        versions_todo = versions_todo[:BATCH_SIZE].generate_work_entries(start.date(), stop.date(), False)
-        # if necessary, retrigger the cron to generate more work entries
-        if version_todo_count > BATCH_SIZE:
-            self.env.ref('hr_work_entry.ir_cron_generate_missing_work_entries')._trigger()

--- a/addons/hr_work_entry/models/hr_work_entry.py
+++ b/None
@@ -1,341 +0,0 @@
-# Part of Odoo. See LICENSE file for full copyright and licensing details.
-
-from collections import defaultdict
-from contextlib import contextmanager
-from datetime import datetime, time, timedelta, UTC
-from itertools import chain
-
-from psycopg2 import OperationalError
-
-from odoo import api, fields, models
-from odoo.exceptions import UserError, ValidationError
-from odoo.fields import Domain
-from odoo.tools import float_compare
-from odoo.tools.intervals import Intervals
-
-
-class HrWorkEntry(models.Model):
-    _name = 'hr.work.entry'
-    _description = 'HR Work Entry'
-    _order = 'create_date'
-
-    name = fields.Char("Description")
-    active = fields.Boolean(default=True)
-    employee_id = fields.Many2one('hr.employee', required=True, domain="['|', ('company_id', '=', False), ('company_id', '=', company_id)]", index=True)
-    version_id = fields.Many2one('hr.version', string="Employee Record", required=True, index=True)
-    work_entry_source = fields.Selection(related='version_id.work_entry_source')
-    resource_calendar_id = fields.Many2one(related='version_id.resource_calendar_id')
-    date = fields.Date(required=True)
-    duration = fields.Float(string="Duration", default=lambda self: self.env.company.resource_calendar_id.hours_per_day)
-    work_entry_type_id = fields.Many2one(
-        'hr.work.entry.type',
-        index=True,
-        default=lambda self: self.env['hr.work.entry.type'].search([], limit=1),
-        domain=lambda self: self._get_work_entry_type_domain())
-    display_code = fields.Char(related='work_entry_type_id.display_code')
-    code = fields.Char(related='work_entry_type_id.code')
-    shortcut_behavior = fields.Selection(related='work_entry_type_id.shortcut_behavior')
-    external_code = fields.Char(related='work_entry_type_id.external_code')
-    color = fields.Integer(related='work_entry_type_id.color', readonly=True)
-    category = fields.Selection(related='work_entry_type_id.category')
-    is_manual = fields.Boolean(help="Marks if the work entry came from manual creation or split")
-    state = fields.Selection([
-        ('draft', 'New'),
-        ('conflict', 'In Conflict'),
-        ('validated', 'In Payslip'),
-        ('cancelled', 'Cancelled')
-    ], default='draft')
-    company_id = fields.Many2one('res.company', string='Company', readonly=True, required=True,
-        default=lambda self: self.env.company)
-    conflict = fields.Boolean('Conflicts', compute='_compute_conflict', store=True)  # Used to show conflicting work entries first
-    department_id = fields.Many2one('hr.department', related='employee_id.department_id', store=True)
-    amount_rate = fields.Float("Pay rate")
-    country_id = fields.Many2one('res.country', related='employee_id.company_id.country_id', search='_search_country_id')
-
-    # FROM 7s by query to 2ms (with 2.6 millions entries)
-    _contract_date_start_stop_idx = models.Index("(version_id, date) WHERE state IN ('draft', 'validated')")
-
-    @api.constrains('duration')
-    def _check_duration(self):
-        for work_entry in self:
-            if float_compare(work_entry.duration, 0, 3) <= 0 or float_compare(work_entry.duration, 24, 3) > 0:
-                raise ValidationError(self.env._("Duration must be positive and cannot exceed 24 hours."))
-
-    @api.depends('display_code', 'duration')
-    def _compute_display_name(self):
-        for work_entry in self:
-            duration = str(timedelta(hours=work_entry.duration)).split(":")
-            work_entry.display_name = "%s - %sh%s" % (work_entry.work_entry_type_id.name, duration[0], duration[1])
-
-    @api.depends('state')
-    def _compute_conflict(self):
-        for rec in self:
-            rec.conflict = rec.state == 'conflict'
-
-    @api.onchange('employee_id', 'date')
-    def _onchange_version_id(self):
-        vals = {
-            'employee_id': self.employee_id.id,
-            'date': self.date,
-        }
-        try:
-            res = self._set_current_contract(vals)
-        except ValidationError:
-            return
-        if version_id := res.get('version_id'):
-            self.version_id = version_id
-
-    @api.model
-    def _set_current_contract(self, vals):
-        if not vals.get('version_id') and vals.get('date') and vals.get('employee_id'):
-            employee = self.env['hr.employee'].browse(vals.get('employee_id'))
-            active_version = employee._get_version(vals['date'])
-            return dict(vals, version_id=active_version.id)
-        return vals
-
-    @api.model
-    def get_unusual_days(self, date_from, date_to=None):
-        return self.env.company.resource_calendar_id._get_unusual_days(
-            datetime.combine(fields.Date.from_string(date_from), time.min, tzinfo=UTC),
-            datetime.combine(fields.Date.from_string(date_to), time.max, tzinfo=UTC),
-            self.company_id,
-        )
-
-    def action_validate(self):
-        """
-        Try to validate work entries.
-        If some errors are found, set `state` to conflict for conflicting work entries
-        and validation fails.
-        :return: True if validation succeeded
-        """
-        work_entries = self.filtered(lambda work_entry: work_entry.state != 'validated')
-        if not work_entries._check_if_error():
-            work_entries.write({'state': 'validated'})
-            return True
-        return False
-
-    def action_split(self, vals):
-        self.ensure_one()
-        if self.duration < 1:
-            raise UserError(self.env._("You can't split a work entry with less than 1 hour."))
-        split_duration = vals['duration']
-        if self.duration <= split_duration:
-            raise UserError(
-                self.env._(
-                    "Split work entry duration has to be less than the existing work entry duration."
-                )
-            )
-        self.duration -= split_duration
-        self.is_manual = True
-        split_work_entry = self.copy()
-        split_work_entry.write(vals)
-        return split_work_entry.id
-
-    def _check_if_error(self):
-        if not self:
-            return False
-        undefined_type = self.filtered(lambda b: not b.work_entry_type_id)
-        undefined_type.write({'state': 'conflict'})
-        conflict = self._mark_conflicting_work_entries()
-        outside_calendar = self._mark_leaves_outside_schedule()
-        return undefined_type or conflict or outside_calendar
-
-    def _mark_conflicting_work_entries(self):
-        """
-        Set `state` to `conflict` for work entries where, for the same employee and day,
-        the total duration exceeds 24 hours.
-        Return True if such entries are found.
-        """
-        start = min(self.mapped('date'))
-        stop = max(self.mapped('date'))
-        self.flush_model(['date', 'duration', 'employee_id', 'active'])
-        query = """
-            WITH excessive_days AS (
-                SELECT employee_id, date
-                FROM hr_work_entry
-                WHERE active = TRUE
-                  AND date BETWEEN %(start)s AND %(stop)s
-                  AND employee_id IN %(employee_ids)s
-                GROUP BY employee_id, date
-                HAVING 0 >= SUM(duration) OR SUM(duration) > 24
-            )
-            SELECT we.id
-            FROM hr_work_entry we
-            JOIN excessive_days ed
-              ON we.employee_id = ed.employee_id
-             AND we.date = ed.date
-            WHERE we.active = TRUE
-        """
-        self.env.cr.execute(query, {
-            "start": start,
-            "stop": stop,
-            'employee_ids': tuple(self.employee_id.ids),
-        })
-        conflict_ids = [row[0] for row in self.env.cr.fetchall()]
-        self.browse(conflict_ids).write({'state': 'conflict'})
-        return bool(conflict_ids)
-
-    def _get_leaves_entries_outside_schedule(self):
-        return self.filtered(lambda w: w.work_entry_type_id.category == 'absence' and w.state not in ('validated', 'cancelled'))
-
-    def _mark_leaves_outside_schedule(self):
-        """
-        Check leave work entries in `self` which are completely outside
-        the contract's theoretical calendar schedule. Mark them as conflicting.
-        :return: leave work entries completely outside the contract's calendar
-        """
-        work_entries = self._get_leaves_entries_outside_schedule()
-        entries_by_calendar = defaultdict(lambda: self.env['hr.work.entry'])
-        for work_entry in work_entries:
-            calendar = work_entry.version_id.resource_calendar_id
-            entries_by_calendar[calendar] |= work_entry
-
-        outside_entries = self.env['hr.work.entry']
-        for calendar, entries in entries_by_calendar.items():
-            if not calendar:
-                continue
-            datetime_start = datetime.combine(min(entries.mapped('date')), time.min)
-            datetime_stop = datetime.combine(max(entries.mapped('date')), time.max)
-
-            if calendar:
-                calendar_intervals = calendar._attendance_intervals_batch(datetime_start.replace(tzinfo=UTC), datetime_stop.replace(tzinfo=UTC))[False]
-            else:
-                calendar_intervals = Intervals([(datetime_start.replace(tzinfo=UTC), datetime_stop.replace(tzinfo=UTC), self.env['resource.calendar.attendance'])])
-            entries_intervals = entries._to_intervals()
-            overlapping_entries = self._from_intervals(entries_intervals & calendar_intervals)
-            outside_entries |= entries - overlapping_entries
-        outside_entries.write({'state': 'conflict'})
-        return bool(outside_entries)
-
-    def _to_intervals(self):
-        return Intervals(
-            ((datetime.combine(w.date, time.min, tzinfo=UTC), datetime.combine(w.date, time.max, tzinfo=UTC), w) for w in self),
-            keep_distinct=True)
-
-    @api.model
-    def _from_intervals(self, intervals):
-        return self.browse(chain.from_iterable(recs.ids for start, end, recs in intervals))
-
-    @api.model_create_multi
-    def create(self, vals_list):
-        vals_list = [self._set_current_contract(vals) for vals in vals_list]
-        company_by_employee_id = {}
-        for vals in vals_list:
-            if (
-                not 'amount_rate' in vals
-                and (work_entry_type_id := vals.get('work_entry_type_id'))
-            ):
-                work_entry_type = self.env['hr.work.entry.type'].browse(work_entry_type_id)
-                vals['amount_rate'] = work_entry_type.amount_rate
-            if vals.get('company_id'):
-                continue
-            if vals['employee_id'] not in company_by_employee_id:
-                employee = self.env['hr.employee'].browse(vals['employee_id'])
-                company_by_employee_id[employee.id] = employee.company_id.id
-            vals['company_id'] = company_by_employee_id[vals['employee_id']]
-        work_entries = super().create(vals_list)
-        work_entries._check_if_error()
-        return work_entries
-
-    def check_modification_allowed(self):
-        """
-        Check that the work entries can be modified.
-        A work entry can't be modified if it's validated and linked to a payslip.
-        :return: True if modification is allowed
-        """
-        return True
-
-    def write(self, vals):
-        skip_check = not bool({'date', 'duration', 'employee_id', 'work_entry_type_id', 'active'} & vals.keys())
-        if 'state' in vals:
-            if vals['state'] == 'draft':
-                vals['active'] = True
-            elif vals['state'] == 'cancelled':
-                vals['active'] = False
-                skip_check &= all(self.mapped(lambda w: w.state != 'conflict'))
-
-        if 'active' in vals:
-            vals['state'] = 'draft' if vals['active'] else 'cancelled'
-
-        employee_ids = self.employee_id.ids
-        if 'employee_id' in vals and vals['employee_id']:
-            employee_ids += [vals['employee_id']]
-        with self._error_checking(skip=skip_check, employee_ids=employee_ids):
-            return super().write(vals)
-
-    @api.ondelete(at_uninstall=False)
-    def _unlink_except_validated_work_entries(self):
-        if any(w.state == 'validated' for w in self):
-            raise UserError(self.env._("This work entry is validated. You can't delete it."))
-
-    def unlink(self):
-        employee_ids = self.employee_id.ids
-        with self._error_checking(employee_ids=employee_ids):
-            return super().unlink()
-
-    def _reset_conflicting_state(self):
-        self.filtered(lambda w: w.state == 'conflict').write({'state': 'draft'})
-
-    @contextmanager
-    def _error_checking(self, start=None, stop=None, skip=False, employee_ids=False):
-        """
-        Context manager used for conflicts checking.
-        When exiting the context manager, conflicts are checked
-        for all work entries within a date range. By default, the start and end dates are
-        computed according to `self` (min and max respectively) but it can be overwritten by providing
-        other values as parameter.
-        :param start: datetime to overwrite the default behaviour
-        :param stop: datetime to overwrite the default behaviour
-        :param skip: If True, no error checking is done
-        """
-        try:
-            skip = skip or self.env.context.get('hr_work_entry_no_check', False)
-            start = start or min(self.mapped('date'), default=False)
-            stop = stop or max(self.mapped('date'), default=False)
-            if not skip and start and stop:
-                domain = (
-                    Domain('date', '<=', stop)
-                    & Domain('date', '>=', start)
-                    & Domain('state', 'not in', ('validated', 'cancelled'))
-                )
-                if employee_ids:
-                    domain &= Domain('employee_id', 'in', list(employee_ids))
-                work_entries = self.sudo().with_context(hr_work_entry_no_check=True).search(domain)
-                work_entries._reset_conflicting_state()
-            yield
-        except OperationalError:
-            # the cursor is dead, do not attempt to use it or we will shadow the root exception
-            # with a "psycopg2.InternalError: current transaction is aborted, ..."
-            skip = True
-            raise
-        finally:
-            if not skip and start and stop:
-                # New work entries are handled in the create method,
-                # no need to reload work entries.
-                work_entries.exists()._check_if_error()
-
-    def _get_work_entry_type_domain(self):
-        if len(self.env.companies.country_id.ids) > 1:
-            return [('country_id', '=', False)]
-        return ['|', ('country_id', '=', False), ('country_id', 'in', self.env.companies.country_id.ids)]
-
-    def _get_source_action_values(self):
-        return {
-            'name': self.env._('Calendar'),
-            'res_model': 'resource.calendar',
-            'res_id': self.resource_calendar_id.id,
-            'views': [(False, 'form')],
-        }
-
-    def action_view_source(self):
-        action = {
-            'type': 'ir.actions.act_window',
-            'target': 'current',
-        }
-        if self.is_manual:
-            raise UserError(self.env._("This work entry was created manually and has no source attendance."))
-        action.update(self._get_source_action_values())
-        return action
-
-    def _search_country_id(self, operator, value):
-        return [('employee_id.company_id.partner_id.country_id', operator, value)]

--- a/addons/hr_work_entry/models/hr_work_entry_type.py
+++ b/addons/hr_work_entry/models/hr_work_entry_type.py
@@ -6,11 +6,15 @@ from odoo.exceptions import UserError
 
 class HrWorkEntryType(models.Model):
     _name = 'hr.work.entry.type'
-    _description = 'HR Work Entry Type'
+    _description = 'Work Entry Type'
+    _order = 'sequence'
 
     name = fields.Char(required=True, translate=True)
     display_code = fields.Char(string="Display Code", size=3, translate=True, help="This code can be changed, it is only for a display purpose (3 letters max)")
-    code = fields.Char(string="Payroll Code", required=True, help="The code is used as a reference in salary rules. Careful, changing an existing code can lead to unwanted behaviors.")
+    code = fields.Char(
+        string="Payroll Code",
+        required=True,
+        help="The code is used as a reference in salary rules. Careful, changing an existing code can lead to unwanted behaviors.")
     external_code = fields.Char(help="Use this code to export your data to a third party")
     color = fields.Integer(default=0)
     sequence = fields.Integer(default=25)
@@ -23,7 +27,7 @@ class HrWorkEntryType(models.Model):
         domain=lambda self: [('id', 'in', self.env.companies.country_id.ids)]
     )
     country_code = fields.Char(related='country_id.code')
-    category = fields.Selection(
+    count_as = fields.Selection(
         [("working_time", "Working Time"), ("absence", "Absence")],
         default="working_time",
         required=True,
@@ -44,24 +48,14 @@ class HrWorkEntryType(models.Model):
              "prompt a duration and will be added to the existing work entries while replace will simply replace all "
              "work entries of that day")
 
-    @api.constrains('country_id')
-    def _check_work_entry_type_country(self):
-        if self.env.ref('hr_work_entry.work_entry_type_attendance') in self:
-            raise UserError(_("You can't change the country of this specific work entry type."))
-        elif not self.env.context.get('install_mode') and self.env['hr.work.entry'].sudo().search_count([('work_entry_type_id', 'in', self.ids)], limit=1):
-            raise UserError(_("You can't change the Country of this work entry type cause it's currently used by the system. You need to delete related working entries first."))
-
     @api.constrains('code', 'country_id')
     def _check_code_unicity(self):
-        similar_work_entry_types = self.search([
+        similar_work_entry_types_by_code = self.search([
             ('code', 'in', self.mapped('code')),
             ('country_id', 'in', self.country_id.ids + [False]),
-            ('id', 'not in', self.ids)
-        ])
-        for work_entry_type in self:
-            invalid_work_entry_types = similar_work_entry_types.filtered_domain([
-                ('code', '=', work_entry_type.code),
-                ('country_id', 'in', self.country_id.ids + [False]),
-            ])
-            if invalid_work_entry_types:
-                raise UserError(_("The same code cannot be associated to multiple work entry types (%s)", ', '.join(list(set(invalid_work_entry_types.mapped('code'))))))
+        ]).grouped('code')
+        for code, similar_work_entry_types in similar_work_entry_types_by_code.items():
+            if len(similar_work_entry_types) == 1:
+                continue
+            if len(similar_work_entry_types) > 2 or (similar_work_entry_types.country_id and len(similar_work_entry_types.country_id.code) != 2):
+                UserError(_("The same code cannot be associated to multiple work entry types for the same company (%s)", code))

--- a/None
+++ b/addons/hr_work_entry/models/res_company.py
@@ -0,0 +1,9 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import fields, models
+
+
+class ResCompany(models.Model):
+    _inherit = 'res.company'
+
+    external_code = fields.Char("External Code", groups="hr.group_hr_user", copy=False, help="Use this code to export your data to a third party")

--- a/None
+++ b/addons/hr_work_entry/models/res_config_settings.py
@@ -0,0 +1,10 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import fields, models
+
+
+class ResConfigSettings(models.TransientModel):
+    _inherit = 'res.config.settings'
+
+    external_code = fields.Char(
+        related='company_id.external_code', string="External Code", readonly=False)

--- a/addons/hr_work_entry/models/resource_calendar.py
+++ b/addons/hr_work_entry/models/resource_calendar.py
@@ -7,9 +7,9 @@ from odoo.fields import Domain
 class ResourceCalendar(models.Model):
     _inherit = 'resource.calendar'
 
-    # Override the method to add 'attendance_ids.work_entry_type_id.category' to
+    # Override the method to add 'attendance_ids.work_entry_type_id.count_as' to
     # the dependencies
-    @api.depends('attendance_ids.work_entry_type_id.category')
+    @api.depends('attendance_ids.work_entry_type_id.count_as')
     def _compute_hours_per_week(self):
         super()._compute_hours_per_week()
 
@@ -19,7 +19,7 @@ class ResourceCalendar(models.Model):
             Domain.OR(
                 [
                     Domain('work_entry_type_id', '=', False),
-                    Domain('work_entry_type_id.category', '=', 'working_time'),
+                    Domain('work_entry_type_id.count_as', '=', 'working_time'),
                 ],
             ),
         )

--- a/addons/hr_work_entry/models/resource_calendar_attendance.py
+++ b/addons/hr_work_entry/models/resource_calendar_attendance.py
@@ -1,17 +1,30 @@
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
-from odoo import models, fields
+from odoo import api, models, fields
 
 
 class ResourceCalendarAttendance(models.Model):
     _inherit = 'resource.calendar.attendance'
 
-    def _default_work_entry_type_id(self):
-        return self.env.ref('hr_work_entry.work_entry_type_attendance', raise_if_not_found=False)
+    work_entry_type_id = fields.Many2one('hr.work.entry.type', 'Work Entry Type', store=True,
+        compute='_compute_work_entry_type_id', groups="hr.group_hr_user")
 
-    work_entry_type_id = fields.Many2one(
-        'hr.work.entry.type', 'Work Entry Type', default=_default_work_entry_type_id,
-        groups="hr.group_hr_user")
+    @api.depends('calendar_id.country_id')
+    def _compute_work_entry_type_id(self):
+        types_per_calendar = {}
+        for resource_calendar_attendence in self:
+            if not resource_calendar_attendence.work_entry_type_id and resource_calendar_attendence.calendar_id:
+                calendar = resource_calendar_attendence.calendar_id
+                if calendar in types_per_calendar:
+                    resource_calendar_attendence.work_entry_type_id = types_per_calendar[calendar]
+                else:
+                    all_types = self.env['hr.work.entry.type'].sudo().search([('code', '=', 'WORK100')])
+                    default_work_entry_type = all_types.filtered(
+                        lambda t: t.country_id == calendar.country_id
+                    ) or all_types.filtered(lambda t: not t.country_id)
+                    if default_work_entry_type:
+                        resource_calendar_attendence.work_entry_type_id = default_work_entry_type[0]
+                        types_per_calendar[calendar] = default_work_entry_type[0]
 
     def _copy_attendance_vals(self):
         res = super()._copy_attendance_vals()
@@ -19,4 +32,4 @@ class ResourceCalendarAttendance(models.Model):
         return res
 
     def _is_work_period(self):
-        return self.work_entry_type_id.category == 'working_time' and super()._is_work_period()
+        return self.work_entry_type_id.count_as == 'working_time' and super()._is_work_period()
